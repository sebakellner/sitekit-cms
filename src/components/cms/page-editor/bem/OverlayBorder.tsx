import React from 'react'
import { Box } from 'grommet'
import {
  BEM_BORDER_COLOR,
  BEM_BORDER_SIZE_SELECTED,
  BEM_BORDER_SIZE_DEFAULT,
  BEM_BORDER_ZINDEX,
} from './constants'

interface OverlayBorderProps {
  isSelected: boolean
  onSelect?: () => void
}

export const OverlayBorder: React.FC<OverlayBorderProps> = ({
  isSelected,
  onSelect,
}) => (
  <Box
    border={{
      color: BEM_BORDER_COLOR,
      size: isSelected ? BEM_BORDER_SIZE_SELECTED : BEM_BORDER_SIZE_DEFAULT,
    }}
    style={{
      position: 'absolute',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
      zIndex: BEM_BORDER_ZINDEX,
      cursor: 'default',
      pointerEvents: onSelect ? 'auto' : 'none',
    }}
    onClick={onSelect}
    hoverIndicator={false}
    focusIndicator={false}
  />
)
