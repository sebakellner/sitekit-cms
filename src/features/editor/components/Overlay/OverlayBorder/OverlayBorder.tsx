import React from 'react'
import {
  BEM_BORDER_COLOR,
  BEM_BORDER_SIZE_DEFAULT,
  BEM_BORDER_ZINDEX,
} from '@src/features/editor/constants/inspector'
import type { OverlayBorderProps } from './OverlayBorder.types'
import { Box } from '@chakra-ui/react'

const OverlayBorder: React.FC<OverlayBorderProps> = ({ onSelect }) => (
  <Box
    borderColor={BEM_BORDER_COLOR}
    borderWidth={BEM_BORDER_SIZE_DEFAULT}
    data-testid="overlay-border"
    style={{
      position: 'absolute',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
      zIndex: BEM_BORDER_ZINDEX,
      cursor: 'default',
      pointerEvents: onSelect ? 'auto' : 'none',
    }}
    onClick={onSelect}
  />
)

export default OverlayBorder
