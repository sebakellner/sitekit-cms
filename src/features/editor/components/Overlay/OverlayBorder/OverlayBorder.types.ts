export interface OverlayBorderProps {
  onSelect?: () => void
}
