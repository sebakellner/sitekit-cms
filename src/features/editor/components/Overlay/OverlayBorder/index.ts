export { default as OverlayBorder } from './OverlayBorder'
export type { OverlayBorderProps } from './OverlayBorder.types'
