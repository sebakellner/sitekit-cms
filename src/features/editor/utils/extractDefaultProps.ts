import { ComponentMetaSchema } from '@components/site/schemas/componentMeta.schema'
import type { ComponentMeta } from '@components/site/types'

function isDefaultProp(val: unknown): val is { default: unknown } {
  return typeof val === 'object' && val !== null && 'default' in val
}

export function extractDefaultProps(
  props: ComponentMeta['props'] | undefined,
  fallback: unknown = undefined
): Record<string, unknown> {
  if (!props) return {}
  const result = ComponentMetaSchema.shape.props.safeParse(props)
  if (!result.success) {
    console.error('Invalid props schema:', result.error)
    return {}
  }
  const validProps = result.data
  return Object.fromEntries(
    Object.entries(validProps).map(([key, val]) => [
      key,
      isDefaultProp(val) ? val.default : fallback,
    ])
  )
}
