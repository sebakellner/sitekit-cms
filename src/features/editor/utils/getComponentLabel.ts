import React from 'react'

export function getComponentLabel(
  children: React.ReactNode,
  fallback = 'Component'
): string {
  if (React.isValidElement(children)) {
    const type = children.type
    if (typeof type === 'function' && type.name) return type.name
  }
  return fallback
}
