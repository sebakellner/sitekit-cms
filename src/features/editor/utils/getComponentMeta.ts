import { ComponentMetaSchema } from '@components/site/schemas/componentMeta.schema'

type ModuleWithDefault = { default: unknown }

export function getComponentMeta(module: unknown) {
  if (
    module &&
    typeof module === 'object' &&
    'default' in module &&
    (module as ModuleWithDefault).default &&
    typeof (module as ModuleWithDefault).default === 'object'
  ) {
    const def = (module as ModuleWithDefault).default
    const result = ComponentMetaSchema.safeParse(def)
    if (result.success) {
      return result.data
    }
  }
  return { name: '', props: {} }
}
