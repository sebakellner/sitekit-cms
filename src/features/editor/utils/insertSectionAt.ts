import { componentsRegistry } from '@registry/componentRegistry'
import { v4 as uuidv4 } from 'uuid'

import type { ComponentMeta } from '@components/site/types'
import { extractDefaultProps } from '@src/features/editor/utils/extractDefaultProps'
import type { Section } from '../schemas/section.schema'

export async function insertSectionAtUtil(
  sections: Section[],
  setSections: (sections: Section[]) => void,
  name: string,
  index: number
) {
  const loader = componentsRegistry[name as keyof typeof componentsRegistry]
  if (!loader) {
    console.warn(`Component "${name}" not found in registry.`)
    return
  }

  const metaModule = await loader()

  const meta = metaModule.default as ComponentMeta

  const defaultProps = extractDefaultProps(meta?.props)

  const newSection: Section = {
    id: uuidv4(),
    name,
    props: defaultProps,
  }

  const updatedSections = [
    ...sections.slice(0, index),
    newSection,
    ...sections.slice(index),
  ]

  setSections(updatedSections)
}
