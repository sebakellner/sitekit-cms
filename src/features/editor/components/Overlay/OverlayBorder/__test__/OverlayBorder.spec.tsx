import { describe, test, expect } from 'vitest'
import { render } from 'test/utils/render'
import type { OverlayBorderProps } from '../OverlayBorder.types'
import OverlayBorder from '../OverlayBorder'

const setup = (overrides: Partial<OverlayBorderProps> = {}) => {
  const defaultProps: OverlayBorderProps = {
    onSelect: vi.fn(),
  }
  const props = { ...defaultProps, ...overrides }

  return <OverlayBorder {...props} />
}

describe('OverlayBorder', () => {
  test('should match with snapshot', () => {
    const { container } = render(setup())
    expect(container).toMatchSnapshot()
  })

  test('should pointer events be auto when onSelect is provided', () => {
    const { getByTestId } = render(setup())
    const overlayBorder = getByTestId('overlay-border')
    expect(overlayBorder).toHaveStyle('pointer-events: auto')
  })

  test('should pointer events be none when onSelect is not provided', () => {
    const { getByTestId } = render(setup({ onSelect: undefined }))
    const overlayBorder = getByTestId('overlay-border')
    expect(overlayBorder).toHaveStyle('pointer-events: none')
  })
})
