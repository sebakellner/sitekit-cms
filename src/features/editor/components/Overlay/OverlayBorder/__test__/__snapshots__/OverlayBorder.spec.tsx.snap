// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OverlayBorder > should match with snapshot 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="css-kwi4bo"
    data-testid="overlay-border"
    style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; z-index: 2; cursor: default; pointer-events: auto;"
  />
</div>
`;
